### ISSUE TITLE
DCP fails to save "initial_lr" field, created by an LR scheduler

### ISSUE THREAD

**Issue by ad8e** [pytorch/pytorch#140900]
### 🐛 Describe the bug

Initial LR is present in the optimizer's state dict on saving, but it disappears on loading. This prevents resuming an LR scheduler; we need to use hacks around `last_epoch=-1` or manually patching `initial_lr` to make one work.

Testcase:
```python
import os

import torch
import torch.distributed as dist
import torch.distributed.checkpoint as dcp
import torch.multiprocessing as mp
import torch.nn as nn

from torch.distributed.fsdp import FullyShardedDataParallel as FSDP
from torch.distributed.checkpoint.state_dict import get_state_dict, set_state_dict
from torch.distributed.checkpoint.stateful import Stateful
from torch.distributed.fsdp.fully_sharded_data_parallel import StateDictType
from torch.distributed.checkpoint import StorageWriter

from torch.distributed.checkpoint.state_dict import (
    get_model_state_dict,
    get_optimizer_state_dict,
    set_model_state_dict,
    set_optimizer_state_dict,
    StateDictOptions
)
CHECKPOINT_DIR = "checkpoint"


class ToyModel(nn.Module):
    def __init__(self):
        super(ToyModel, self).__init__()
        self.net2 = nn.Linear(1, 1, bias=False)

    def forward(self, x):
        return self.net2(x)

class ModelWrapper(Stateful):
    def __init__(self, model: nn.Module) -> None:
        self.model = model

    def state_dict(self) -> None:
        return get_model_state_dict(self.model)

    def load_state_dict(self, state_dict) -> None:
        set_model_state_dict(self.model, state_dict)


class OptimizerWrapper(Stateful):
    def __init__(self, model: nn.Module, optim: torch.optim.Optimizer) -> None:
        self.model = model
        self.optim = optim

    def state_dict(self) -> None:
        return get_optimizer_state_dict(self.model, self.optim, options=StateDictOptions(flatten_optimizer_state_dict=True))

    def load_state_dict(self, state_dict) -> None:
        set_optimizer_state_dict(self.model, self.optim, optim_state_dict=state_dict, options=StateDictOptions(flatten_optimizer_state_dict=True))

if __name__ == "__main__":
    rank = 0
    world_size = 1
    os.environ["MASTER_ADDR"] = "localhost"
    os.environ["MASTER_PORT"] = "12355 "

    dist.init_process_group("cuda:nccl", rank=rank, world_size=world_size)
    torch.cuda.set_device(rank)

    model = ToyModel().to(rank)

    optimizer = torch.optim.Adam(model.parameters(), lr=0.1)
    lr_scheduler = torch.optim.lr_scheduler.StepLR(optimizer, step_size=1, gamma=0.1)

    loss = model(torch.rand(1, device="cuda")).sum()
    loss.backward()
    optimizer.step()
    lr_scheduler.step()
    optimizer.zero_grad()  # uncomment me to fix DCP!

    state_dict = {"model": ModelWrapper(model), "optimizer": OptimizerWrapper(model, optimizer)}
    dcp.save(state_dict, checkpoint_id=f"{CHECKPOINT_DIR}")

    with torch.no_grad():
        print("optimizer", optimizer.state_dict())
        optimizer = torch.optim.Adam(model.parameters(), lr=20)
        model.net2.weight.zero_()
        print("optimizer is cleared:", optimizer.state_dict())
        print("model is cleared:", model.state_dict())
        state_dict = {"model": ModelWrapper(model), "optimizer": OptimizerWrapper(model, optimizer)}
        dcp.load(state_dict, checkpoint_id=f"{CHECKPOINT_DIR}")
        print("optimizer after load:", optimizer.state_dict())
        print("model after load:", model.net2.weight)
        dist.destroy_process_group()
```
(same code as issue #140898)

Output:
```
optimizer {'state': {0: {'step': tensor(1.), 'exp_avg': tensor([[0.0834]], device='cuda:0'), 'exp_avg_sq': tensor([[0.0007]], device='cuda:0')}}, 'param_groups': [{'lr': 0.010000000000000002, 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0, 'amsgrad': False, 'maximize': False, 'foreach': None, 'capturable': False, 'differentiable': False, 'fused': None, 'initial_lr': 0.1, 'params': [0]}]}
...
optimizer after load: {'state': {0: {'step': tensor(1.), 'exp_avg': tensor([[0.0834]], device='cuda:0'), 'exp_avg_sq': tensor([[0.0007]], device='cuda:0')}}, 'param_groups': [{'lr': 0.010000000000000002, 'betas': (0.9, 0.999), 'eps': 1e-08, 'weight_decay': 0, 'amsgrad': False, 'maximize': False, 'foreach': None, 'capturable': False, 'differentiable': False, 'fused': None, 'params': [0]}]}
```
Notice that the initial_lr field has disappeared.

In my recollection, any other field added to the `param_groups` section will likewise disappear after saving and loading.

### Versions

```
PyTorch version: 2.5.0
Is debug build: False
CUDA used to build PyTorch: 12.6
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.5 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.30.5
Libc version: glibc-2.35

Python version: 3.10.12 (main, Sep 11 2024, 15:47:36) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.5.13-650-3434-22042-coreweave-1-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: 12.6.77
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: GPU 0: NVIDIA H100 80GB HBM3
Nvidia driver version: 535.216.01
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_adv.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_cnn.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_precompiled.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_engines_runtime_compiled.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_graph.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_heuristic.so.9.3.0
/usr/lib/x86_64-linux-gnu/libcudnn_ops.so.9.3.0
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 57 bits virtual
Byte Order:                         Little Endian
CPU(s):                             128
On-line CPU(s) list:                0-127
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Platinum 8462Y+
CPU family:                         6
Model:                              143
Thread(s) per core:                 2
Core(s) per socket:                 32
Socket(s):                          2
Stepping:                           8
CPU max MHz:                        4100.0000
CPU min MHz:                        800.0000
BogoMIPS:                           5600.00
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf tsc_known_freq pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cat_l2 cdp_l3 invpcid_single cdp_l2 ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm rdt_a avx512f avx512dq rdseed adx smap avx512ifma clflushopt clwb intel_pt avx512cd sha_ni avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local split_lock_detect avx_vnni avx512_bf16 wbnoinvd dtherm ida arat pln pts hfi vnmi avx512vbmi umip pku ospke waitpkg avx512_vbmi2 gfni vaes vpclmulqdq avx512_vnni avx512_bitalg tme avx512_vpopcntdq la57 rdpid bus_lock_detect cldemote movdiri movdir64b enqcmd fsrm md_clear serialize tsxldtrk pconfig arch_lbr ibt amx_bf16 avx512_fp16 amx_tile amx_int8 flush_l1d arch_capabilities
Virtualization:                     VT-x
L1d cache:                          3 MiB (64 instances)
L1i cache:                          2 MiB (64 instances)
L2 cache:                           128 MiB (64 instances)
L3 cache:                           120 MiB (2 instances)
NUMA node(s):                       2
NUMA node0 CPU(s):                  0,2,4,6,8,10,12,14,16,18,20,22,24,26,28,30,32,34,36,38,40,42,44,46,48,50,52,54,56,58,60,62,64,66,68,70,72,74,76,78,80,82,84,86,88,90,92,94,96,98,100,102,104,106,108,110,112,114,116,118,120,122,124,126
NUMA node1 CPU(s):                  1,3,5,7,9,11,13,15,17,19,21,23,25,27,29,31,33,35,37,39,41,43,45,47,49,51,53,55,57,59,61,63,65,67,69,71,73,75,77,79,81,83,85,87,89,91,93,95,97,99,101,103,105,107,109,111,113,115,117,119,121,123,125,127
Vulnerability Gather data sampling: Not affected
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Not affected
Vulnerability Retbleed:             Not affected
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced / Automatic IBRS; IBPB conditional; RSB filling; PBRSB-eIBRS SW sequence; BHI BHI_DIS_S
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Not affected

Versions of relevant libraries:
[pip3] clip-anytorch==2.6.0
[pip3] dctorch==0.1.2
[pip3] DISTS-pytorch==0.1
[pip3] gpytorch==1.13
[pip3] lovely-numpy==0.2.13
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.4
[pip3] torch==2.5.0
[pip3] torchaudio==2.5.0
[pip3] torchdiffeq==0.2.4
[pip3] torchsde==0.2.6
[pip3] torchvision==0.20.0
[pip3] triton==3.1.0
[pip3] welford-torch==0.2.4
[conda] Could not collect
```

cc @H-Huang @awgu @kwen2501 @wanchaol @fegin @fduwjj @wz337 @wconstab @d4l3k @c-p-i-o

**Comment by ad8e** [comment]
Note that I have not tested regular torch.save here. This behavior may be a property of the optimizer rather than DCP.

**Comment by fegin** [comment]
There is no easy way to solve this problem because DCP requires states to be present before loading. `set_optim_state_dict()` does its best to initialize the optimizer states but it does not know the existence of `lr_scheduler`. Hence the missing `initial_lr`.

**Comment by ad8e** [comment]
The `lr_scheduler` that creates the `initial_lr` field in the optimizer's param groups does not need to be saved; it can be recreated from scratch. But the `initial_lr` field is already in the optimizer state dict, so why does it disappear while `lr` is kept? It seems like both of them behave the same way.

**Comment by fegin** [comment]
It's saved in the checkpoint. But because when you create a new `lr_scheduler` with `last_epoch != -1`, the `lr_scheduler` won't insert `initial_lr` into `optim.param_groups`. Based on this information, DCP will not load the saved `initial_lr`. On the other hand, when you create a optimizer, `lr` is automatically created in `optim.param_groups`.

**Comment by ad8e** [comment]
Understood! So to avoid this issue, the user can initialize the LR scheduler with `last_epoch = -1` or manually create a dummy `initial_lr` before doing the DCP load. Since the data can be extracted with the right technique, docs would fix this issue.

**Comment by fegin** [comment]
I think setting `last_epoch = -1` is the right solution. In principal, `lr_scheduler` would also be saved in the `state_dict`. So `last_epoch` will also be restored correctly. Again, will keep this issue open until we fix the document.

**Comment by ad8e** [comment]
A comment about API for LR schedules and LRs: it's very common for users to change these halfway in a run, as loss spikes appear or training stop times are chosen. In these cases, the user wants the LR in the current config (which is present in the optimizer before DCP loading happens) over the LR loaded from the checkpoint.

A deficiency of Pytorch's `LRScheduler` API here is that it creates initial_LR based on the current LR. So if you create the LR schedule before DCP loading, the checkpoint's LR wins. And if you create the LR schedule (with `last_epoch=-1`) after DCP loading, then it sees the already-decayed LR and creates a wrong `initial_lr`.

In reality, the ideal solution would be to load only `state` but not `param_groups` from the checkpoint, but I'm working with the current situation, haha. `param_groups` are set by config and should take priority over the checkpoint, but `state` is not in the config.

**Comment by fegin** [comment]
Basically, if optim.state_dict() also returns `param_groups`, then DCP has to save it. Otherwise it is a surprise to most users. I admit that optimizer and lr_scheduler state_dict is not very well-structure, especially supporting distributed checkpointing. Given the possibility of BC, it is unlikely to change how optim and lr_scheduler state_dict. So if users do want to have some special ways (or common ways as you mentioned it is quite common) that DCP cannot support, I'm afraid users will need to have some workaround.
